lib/linked_list.rb

```diff
@@ -18,130 +18,363 @@ def initialize
+        return
+      end
+
+      current = @head
+      while current != nil
+        if current.next.data == value
+         current.next = current.next.next
+         return
+        end
+
+        current = current.next
+      end
     end
+    #   if @head.nil?
+    #     return nil
+    #   end
+    #   previous = nil
+    #   current = @head
+
+    #   while current != nil
+
+    #     if current.data == value
+    #      previous = current.next 
+    #     end
+
+    #     previous = current
+    #     current = current.next
+    #   end
+    # end
 
     # method to reverse the singly linked list
     # note: the nodes should be moved and not just the values in the nodes
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: 0(n) ?
+    # Space Complexity: o(1)?
     def reverse
-      raise NotImplementedError
-    end
+      if @head.nil?
+       return nil
+     end
+
+     current = @head
+     previous = nil
+     next_node = nil
+
+     while current != nil
+       next_node = current.next
+       current.next = previous
+       previous = current
+       current = next_node
+     end
+     @head = previous
+   end
 
 
     ## Advanced Exercises
     # returns the value at the middle element in the singly linked list
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: 0(n) ?
+    # Space Complexity: 0(1) ?
     def find_middle_value
-      raise NotImplementedError
-    end
+      if @head.nil?
+        return nil
+      end
+      counter = 0
+      current = @head
+
+      while current != nil
+        counter += 1
+        current = current.next
+      end
+
+      mid_index = counter/2
+      node_index = 0
+      current = @head
+
+      while node_index < mid_index
+        node_index += 1
+        current = current.next
+      end
+    return current.data 
+  end
 
     # find the nth node from the end and return its value
     # assume indexing starts at 0 while counting to n
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: 0(n) ?
+    # Space Complexity: 0(1) ?
     def find_nth_from_end(n)
-      raise NotImplementedError
+      if @head.nil?
+        return nil
+      end
+      
+      counter = - 1
+      current = @head
+
+      while current != nil
+        counter += 1
+        current = current.next
+      end
+
+      return nil if n > counter 
+
+      current = @head
+      n_node = counter - n
+      count = 0
+  
+      while current != nil
+        if count == n_node
+         return current.data
+        end
+        
+       count += 1
+       current = current.next
+      end
+      
     end
 
     # checks if the linked list has a cycle. A cycle exists if any node in the
     # linked list links to a node already visited.
     # returns true if a cycle is found, false otherwise.
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: 0(n)?
+    # Space Complexity: 0(n) ?
     def has_cycle

```

Comment:
Clever use of a hash to detect the cycle.  I hope you got the other way with the two node pointers, one fast and one slow we talked about in class.